cli/src/commands/workspace.rs

```diff
@@ -306,22 +360,36 @@ fn cmd_workspace_update_stale(
     // operation, then merge the concurrent operations. The wc_commit_id of the
     // merged repo wouldn't change because the old one wins, but it's probably
     // fine if we picked the new wc_commit_id.
-    let known_wc_commit = {
-        let mut workspace_command = for_stale_working_copy(command, ui)?;
+    let (known_wc_commit, recovered) = {
+        let (mut workspace_command, recovered) = for_stale_working_copy(command, ui)?;
         workspace_command.maybe_snapshot(ui)?;
+
         let wc_commit_id = workspace_command.get_wc_commit_id().unwrap();
-        workspace_command.repo().store().get_commit(wc_commit_id)?
+        (
+            workspace_command.repo().store().get_commit(wc_commit_id)?,
+            recovered,
+        )
     };
     let mut workspace_command = command.workspace_helper_no_snapshot(ui)?;
 
     let repo = workspace_command.repo().clone();
     let (mut locked_ws, desired_wc_commit) =
         workspace_command.unchecked_start_working_copy_mutation()?;
     match check_stale_working_copy(locked_ws.locked_wc(), &desired_wc_commit, &repo)? {
-        WorkingCopyFreshness::Fresh | WorkingCopyFreshness::Updated(_) => writeln!(
-            ui.stderr(),
-            "Nothing to do (the working copy is not stale)."
-        )?,
+        WorkingCopyFreshness::Fresh | WorkingCopyFreshness::Updated(_) => {
+            if !recovered {
```

Comment:
Ah, okay, so the recovery should have already been done if the error was `MissingOperation`. Maybe better to return early in that case?

Reply:
Right now nothing happens after this `match`, so returning early has no effect.

I still don't want to return early, though. If we add code after this `match` (later), it would most likely need to take effect both in this recovery case and in other recovery cases (e.g. the working copy being stale).

Reply:
Then, the code can be structured as follows?
```rust
if !recovered {
    if check_stale_working_copy(...) {
    } else {
         locked_ws
            .locked_wc()
            .check_out(&desired_wc_commit)
            ...
    }
}
```

If I understand it, `recovered = true` means the `.check_out()` has been done.

Reply:
You are probably right that there is some redundancy, but I was thinking that it's better if each recovery step is as independent as possible. This is a recovery process so I think that predictability is much more important than efficiency.

Reply:
Well, the original recovery method is basically as follows:

1. Snapshot the current working copy changes on top of the "stale" operation
2. Merge diverged operations (which will rebase the new snapshot accordingly)
3. Now we're safe to modify the working copy, so update the working copy to the operation head's

(Beware that the step (3) depends on (1).)

The one implemented by this PR is:

1. Create new empty working copy commit for recovery
2. Reset the working copy to it (= the operation head's)
3. (Might be better to snapshot the working copy changes within this operation, but that will be done by the next `jj` invocation)

If I understand it, the recovery should have completed at the step (2), so it doesn't make sense to continue to the step (3) of the original method.

Reply:
> the recovery should have completed at the step (2), so it doesn't make sense to continue to the step (3) of the original method
Conceptually, I think that just because a recovery step was done doesn't mean that there is no longer anything to recover from (there might be more problems). It is true that in this case, after we recover, it is impossible for the working copy to be stale, but I don't think it's true in the general case.

Also, if we later add more recovery steps to `workspace update-stale`, I think it's clearer if each recovery step is allowed to run, rather than try to figure out which ones become redundant due to what happened earlier.

I think we need another tiebreaker vote on this (probably from @martinvonz ).

Reply:
I checked with @martinvonz and I'll go with what @yuja says.

Reply:
Instead of putting a lot of the function in a big `if !recovered` block, I checked `recovered` and exit early if that's the case, which I think is equivalent functionality-wise.